msal/src/main/java/com/microsoft/identity/client/internal/controllers/BrokerMsalController.java

```diff
@@ -188,4 +198,48 @@ private AcquireTokenResult getAcquireTokenResult(final Bundle resultBundle) thro
         throw resultAdapter.baseExceptionFromBundle(resultBundle);
     }
 
+    /**
+     * Checks if the account returns is a MSA Account and sets single on state in cache
+     * @param resultBundle
+     * @param msalOAuth2TokenCache
+     */
+    private void saveMsaAccountToCache(@NonNull final Bundle resultBundle,
+                                       @NonNull final MsalOAuth2TokenCache msalOAuth2TokenCache) {
+        final BrokerResult brokerResult = (BrokerResult) resultBundle.getSerializable(
+                AuthenticationConstants.Broker.BROKER_RESULT_V2
+        );
+
+        if (resultBundle.getBoolean(AuthenticationConstants.Broker.BROKER_REQUEST_V2_SUCCESS)
+                && brokerResult != null &&
+                AzureActiveDirectoryAudience.MSA_MEGA_TENANT_ID.equalsIgnoreCase(brokerResult.getTenantId())) {
+            Logger.info(TAG, "Result returned for MSA Account, saving to cache");
+
+            try {
+                final ClientInfo clientInfo = new ClientInfo(brokerResult.getClientInfo());
+                final MicrosoftStsAccount microsoftStsAccount = new MicrosoftStsAccount(
+                        new IDToken(brokerResult.getIdToken()),
+                        clientInfo
+                );
+                microsoftStsAccount.setEnvironment(brokerResult.getEnvironment());
+
+                final MicrosoftRefreshToken microsoftRefreshToken = new MicrosoftRefreshToken(
+                        brokerResult.getRefreshToken(),
+                        clientInfo,
+                        brokerResult.getScope(),
+                        brokerResult.getClientId(),
+                        brokerResult.getEnvironment(),
+                        brokerResult.getFamilyId()
+                );
+
+                msalOAuth2TokenCache.setSingleSignOnState(microsoftStsAccount, microsoftRefreshToken);
```

Comment:
May be a good idea to log the result value here - should be a true/false on success. Could log a warning on non-success.

Reply:
Throwing back the exception to user